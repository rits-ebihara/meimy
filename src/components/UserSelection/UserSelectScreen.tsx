import {
    Body,
    Button,
    Content,
    Form,
    Icon,
    Input,
    Item,
    List,
    ListItem,
    Right,
    Spinner,
    Text,
    Toast,
    View,
} from 'native-base';
import React, { Component } from 'react';
import { Modal, Picker, TextStyle, ViewStyle } from 'react-native';

import { getConfig } from '../../account-manager/Config';
import { getEimAccount } from '../../account-manager/EimAccount';
import { getDocListValue } from '../../eim-service/EIMServiceAdapter';
import { IDocListRowForView } from '../../eim-service/IDocListForView';
import { IDocListSearchOption, SearchCondition } from '../../eim-service/IDocListSearchOption';
import { DirectoryTypeKey } from '../UserBadge';
import { IGroupList, IOrganizationList, IUserList } from './IUserList';
import { langProfile } from '../../LangProfile';

const config = getConfig();

type DocListName = 'userdoclist' | 'groupdoclist' | 'organizationdoclist';

type directoryTypeName = 'LK_user' | 'LK_organization' | 'LK_group';

const directoryType: { [key in DirectoryTypeKey]: directoryTypeName } = {
    user: 'LK_user',
    group: 'LK_group',
    organization: 'LK_organization',
};

export interface IFilter {
    users: boolean;
    groups: boolean;
    organizations: boolean;
}

interface IOptionalProps {
    filter: IFilter;
}

interface IProps extends Partial<IOptionalProps> {
    onSelect: (docId: string, type: DirectoryTypeKey) => void;
}


interface ISearchedListItem {
    docId: string;
    displayName: string;
    corpName: string;
    orgName: string;
    faceImageId: string;
}

interface IState {
    canContinue: boolean;
    selectedDirectoryType: DirectoryTypeKey;
    searchResult: ISearchedListItem[];
    searchWords: string;
    shown: boolean;
    showNoResultMessage: boolean;
    processing: boolean;
}

const containerStyle: ViewStyle = {
    backgroundColor: '#fff',
    width: '95%',
    height: '95%',
};

const searchBox: ViewStyle = {
    flexDirection: 'row',
    flexGrow: 0,
};

const listOrgStyle: TextStyle = {
    fontSize: 12,
    color: config.colorPalets.$frontDisabledColor,
};

const warnMessageStyle: TextStyle = {
    color: '#999',
    textAlign: 'center',
};
export class UserSelectScreen extends Component<IProps, IState> {
    public static defaultProps: IOptionalProps = {
        filter: {
            users: true,
            groups: true,
            organizations: true,
        }
    }
    private searchedDirType: DirectoryTypeKey = 'user';
    private searchedWord: string = '';
    private searchCondition: {
        offset: number;
        limit: number;
    } = { offset: 0, limit: 30 };
    private getInitState = (props: IProps): IState => ({
        canContinue: false,
        searchResult: [],
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        selectedDirectoryType: this.initSelectedDirType(props.filter!),
        searchWords: '',
        shown: false,
        showNoResultMessage: false,
        processing: false,
    });
    private initSelectedDirType = (filter: IFilter): DirectoryTypeKey => {
        let ret: DirectoryTypeKey = 'user';
        if (filter.groups) {
            ret = 'group'
        }
        if (filter.organizations) {
            ret = 'organization'
        }
        if (filter.users) {
            ret = 'user'
        }
        return ret;
    }
    public constructor(props: IProps) {
        super(props);
        this.state = this.getInitState(props);
    }
    public render() {
        return (
            <Modal transparent animated
                animationType="fade" visible={this.state.shown}>
                <View style={{
                    flex: 1,
                    backgroundColor: 'rgba(0,0,0,0.5)',
                    alignItems: 'center',
                    justifyContent: 'center'
                }}>
                    <View style={containerStyle}>
                        <View style={{ flexDirection: 'row-reverse' }}>
                            <Button key="close-button" icon transparent
                                onPress={this.closeButtonPress}>
                                <Icon name="close" />
                            </Button>
                        </View>
                        <View style={{ flex: 1 }}>
                            <Form style={searchBox}>
                                <Item fixedLabel style={{ flexGrow: 1 }}>
                                    <Input key="search-word-input" value={this.state.searchWords}
                                        placeholder={langProfile.replaceLang('LK_MSG_searchWord')}
                                        onChangeText={this.changeSearchWords}
                                        onSubmitEditing={this.pressSearchButton}
                                        returnKeyType="search" />
                                </Item>
                                <Button key="search-button"
                                    icon transparent style={{ flexGrow: 0 }}
                                    onPress={this.pressSearchButton}>
                                    <Icon name="search" />
                                </Button>
                                <View style={{ flexGrow: 1 }}>
                                    {this.createDirectoryTypePicker(this.state.selectedDirectoryType)}
                                </View>
                            </Form>
                            <Content style={{ flexGrow: 1 }}>
                                {this.state.showNoResultMessage ?
                                    <Text style={warnMessageStyle}>
                                        {langProfile.replaceLang('LK_MSG_applicable')}
                                    </Text> : null}
                                <List key="result-list">
                                    {this.createSearchedUserList()}
                                </List>
                                {/* 更に表示ボタン */}
                                {this.createContinueButton()}
                            </Content>
                        </View>
                    </View>
                </View>
            </Modal>
        );
    }
    public show = () => {
        const s = this.getInitState(this.props);
        s.shown = true;
        this.searchCondition.offset = 0;
        this.setState(s);
    }
    private createContinueButton = () => {
        return (this.state.processing) ?
            <Spinner /> :
            (0 < this.state.searchResult.length && this.state.canContinue) ?
                <Button key="more-search-button" full
                    onPress={this.pressMoreSearch}>
                    <Text>{langProfile.replaceLang('LK_show')}</Text>
                </Button> :
                null
    }
    private closeButtonPress = () => {
        this.setState({
            shown: false,
        });
    }
    private changeSearchWords = (value: string) => {
        this.setState({
            searchWords: value,
        });
    }
    private createSearchedUserList = () => {
        return this.state.searchResult.map(item => (
            <ListItem key={item.docId}>
                <Body>
                    <Text>{item.displayName}</Text>
                    <Text style={listOrgStyle}>{item.orgName}</Text>
                </Body>
                <Right>
                    <Button icon transparent success
                        onPress={this.pressResultRow.bind(this, item.docId)}>
                        <Icon name="md-add-circle" />
                    </Button>
                </Right>
            </ListItem>
        ));
    }
    private pressResultRow = (docId: string) => {
        this.props.onSelect(docId, this.searchedDirType);
        this.closeButtonPress();
    }
    private createDirectoryTypePicker = (selectedDirectoryType: DirectoryTypeKey) => {
        const items: JSX.Element[] = [];
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const filter = this.props.filter!;
        // ユーザー
        if (filter.users) {
            items.push(
                <Picker.Item
                    key={'user'}
                    label={langProfile.replaceLang(directoryType['user'])}
                    value={'user'} />
            );
        }
        if (filter.groups) {
            items.push(
                <Picker.Item
                    key={'group'}
                    label={langProfile.replaceLang(directoryType['group'])}
                    value={'group'} />
            );
        }
        if (filter.organizations) {
            items.push(
                <Picker.Item
                    key={'organization'}
                    label={langProfile.replaceLang(directoryType['organization'])}
                    value={'organization'} />
            );
        }
        if (items.length < 2) {
            return null;
        }
        return (
            <Picker
                key="dir-picker"
                mode="dropdown"
                selectedValue={selectedDirectoryType}
                onValueChange={this.changeDirectoryType.bind(this)}>
                {items}
            </Picker>);
    }

    private changeDirectoryType = (key: DirectoryTypeKey) => {
        this.setState({
            selectedDirectoryType: key,
        });
    }
    private pressMoreSearch = () => {
        this.setState({
            processing: true,
        });
        this.startSearch[this.state.selectedDirectoryType]();
    }
    private pressSearchButton = () => {
        if (this.state.searchWords.trim() === '') { return; }
        this.searchedWord = this.state.searchWords.trim();
        this.searchCondition.offset = 0;
        this.setState({
            canContinue: false,
            searchResult: [],
            processing: true,
            showNoResultMessage: false,
        });
        this.startSearch[this.state.selectedDirectoryType]();
    }
    private startSearch: { [key in DirectoryTypeKey]: () => Promise<void> } = {
        group: () => {
            const search = this.createSearchCondition<IGroupList>([{
                ignoreCaseSense: false,
                operator: 'equal',
                propertyName: 'properties.groupType',
                type: 'string',
                value: 'group',
            }]);
            const condition: IDocListSearchOption<IGroupList> = {
                ...this.searchCondition,
                search,
            };
            return this.commonSearch('groupdoclist',
                this.createGroupRowData, condition);
        },
        organization: () => {
            const search = this.createSearchCondition<IGroupList>();
            const condition: IDocListSearchOption<IGroupList> = {
                ...this.searchCondition,
                search,
            };
            return this.commonSearch('organizationdoclist',
                this.createOrganizationRowData, condition);
        },
        user: () => {
            const search = this.createSearchCondition<IUserList>([
                '(',
                {
                    propertyName: 'properties.userType',
                    type: 'string',
                    operator: 'equal',
                    value: 'user',
                    ignoreCaseSense: false
                },
                'or',
                {
                    propertyName: 'properties.userType',
                    type: 'string',
                    operator: 'equal',
                    value: null,
                    ignoreCaseSense: false
                },
                ')',]);
            const condition: IDocListSearchOption<IUserList> = {
                ...this.searchCondition,
                search,
            };
            return this.commonSearch('userdoclist',
                this.createUserRowData,
                condition);
        }
    }
    private createSearchCondition = <T extends {}>(searches?: SearchCondition<T>[]): SearchCondition<T>[] => {
        const argConditons: SearchCondition<T>[] = !searches ? [] : ['(', ...searches, ')', 'and'];
        const fulltext: SearchCondition<T> = {
            ignoreCaseSense: true,
            propertyName: 'fulltextsearch_list',
            type: 'string',
            operator: 'fulltextsearch_sentence',
            value: this.searchedWord,
        };
        return [...argConditons, fulltext];
    }
    private commonSearch = async <T extends {}>(listName: DocListName,
        createRow: (item: IDocListRowForView<T>) => ISearchedListItem,
        condition: IDocListSearchOption<T>) => {
        this.searchedDirType = this.state.selectedDirectoryType;
        const account = getEimAccount();
        const sa = account.getServiceAdapter();
        try {
            const result = await sa.getDocListForView<T>(
                account.eimTokens,
                'addressbook',
                listName,
                {
                    ...condition,
                });
            const addList = result.docList.map(createRow);
            const newList = this.state.searchResult.concat(addList);
            const canContinue = newList.length < result.metrics.totalCount;
            this.searchCondition.offset += result.docList.length;
            this.setState({
                canContinue,
                searchResult: newList,
                showNoResultMessage: newList.length === 0,
                processing: false,
            });
        } catch {
            Toast.show({
                text: langProfile.replaceLang('LK_netWork'),
                type: 'danger',
            });
        }
    }
    private createUserRowData = (item: IDocListRowForView<IUserList>): ISearchedListItem => {
        const { columnValues } = item;
        return {
            corpName: (getDocListValue<IUserList>(columnValues, 'properties.companyName') as string) || '',
            displayName: (getDocListValue<IUserList>(columnValues, 'properties.displayName') as string) || '',
            faceImageId: '',
            docId: item.documentId,
            orgName: (getDocListValue<IUserList>(columnValues, 'properties.fullLabel') as string) || '',
        };
    }
    private createGroupRowData = (item: IDocListRowForView<IGroupList>): ISearchedListItem => {
        const { columnValues } = item;
        return {
            corpName: '',
            displayName: (getDocListValue<IGroupList>(columnValues, 'properties.fullLabel') as string) || '',
            faceImageId: '',
            docId: item.documentId,
            orgName: '',
        };
    }
    private createOrganizationRowData = (item: IDocListRowForView<IOrganizationList>): ISearchedListItem => {
        const { columnValues } = item;
        return {
            corpName: '',
            displayName: (getDocListValue<IOrganizationList>(columnValues, 'properties.fullLabel') as string) || '',
            faceImageId: '',
            docId: item.documentId,
            orgName: '',
        };
    }
}

export default UserSelectScreen
