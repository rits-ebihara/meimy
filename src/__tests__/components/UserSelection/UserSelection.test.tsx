import Enzyme from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import React from 'react';

import UserSelection, { IUserListItem } from '../../../components/UserSelection/UserSelection';

Enzyme.configure({
    adapter: new EnzymeAdapter(),
});

const getUsers = (): IUserListItem[] => {
    return [
        {
            type: 'user',
            userDocId: 'a0001',
        },
        {
            type: 'group',
            userDocId: 'a0002',
        },
    ];
};

describe('init', () => {
    test('minimal', () => {
        const users = getUsers();
        const target = Enzyme.shallow(
            <UserSelection selectedUsers={users}
                editable multiple
                onChange={jest.fn()}
            />);
        expect(toJson(target)).toMatchSnapshot();
    });
    test('minimal - no multiple', () => {
        const users = getUsers();
        const target = Enzyme.shallow(
            <UserSelection selectedUsers={users}
                editable
                onChange={jest.fn()}
            />);
        expect(toJson(target)).toMatchSnapshot();
    });
    test('user badge styled', () => {
        const users = getUsers();
        const target = Enzyme.shallow(
            <UserSelection
                userBadgeProp={{
                    badgeColor: '#900',
                    textColor: '#ccc',
                    onLongPress: () => { },
                    style: {
                        borderColor: '#090',
                    },
                    type: 'organization',
                }}
                multiple
                selectedUsers={users}
                editable={false}
                onChange={jest.fn()}
            />);
        expect(toJson(target)).toMatchSnapshot();
    });
});

describe('event', () => {
    test('add user', () => {
        const users = getUsers();
        const fn = jest.fn();
        const target = Enzyme.shallow<UserSelection>(
            <UserSelection selectedUsers={users}
                editable multiple
                onChange={fn}
            />);
        const instance = target.instance();
        instance['addList']('add-user-id', 'user');
        expect(fn).toBeCalledWith([...users, {
            type: 'user', userDocId: 'add-user-id'
        }]);
    });
    test('add user - no multiple', () => {
        const users = getUsers();
        const fn = jest.fn();
        const target = Enzyme.shallow<UserSelection>(
            <UserSelection selectedUsers={users}
                editable
                onChange={fn}
            />);
        const instance = target.instance();
        instance['addList']('add-user-id', 'user');
        expect(fn).toBeCalledWith([{
            type: 'user', userDocId: 'add-user-id'
        }]);
    });
    test('add exist user', () => {
        const users = getUsers();
        const fn = jest.fn();
        const target = Enzyme.shallow<UserSelection>(
            <UserSelection selectedUsers={users}
                editable multiple
                onChange={fn}
            />);
        const instance = target.instance();
        instance['addList']('a0001', 'user');
        // onChange のイベントが呼ばれないこと
        expect(fn).not.toBeCalled();
    });
    test('remove user', () => {
        const users = getUsers();
        const fn = jest.fn();
        const target = Enzyme.shallow<UserSelection>(
            <UserSelection selectedUsers={users}
                editable multiple
                onChange={fn}
            />);
        const instance = target.instance();
        instance['onDelete']('a0001');
        expect(fn).toBeCalledWith([{
            type: 'group',
            userDocId: 'a0002',
        }]);
    });
    test('remove user - no editable', () => {
        const users = getUsers();
        const fn = jest.fn();
        const target = Enzyme.shallow<UserSelection>(
            <UserSelection selectedUsers={users}
                multiple
                onChange={fn}
            />);
        const instance = target.instance();
        instance['onDelete']('a0001');
        expect(fn).not.toBeCalled();
    });
    test('remove user - no exist user', () => {
        const users = getUsers();
        const fn = jest.fn();
        const target = Enzyme.shallow<UserSelection>(
            <UserSelection selectedUsers={users}
                editable multiple
                onChange={fn}
            />);
        const instance = target.instance();
        instance['onDelete']('a009');
        expect(fn).not.toBeCalled();
    });
    test('press add user button', () => {
        const users = getUsers();
        const fn = jest.fn();
        const target = Enzyme.shallow<UserSelection>(
            <UserSelection selectedUsers={users}
                editable multiple
                onChange={fn}
            />);
        const instance = target.instance();
        instance['addButtonPress'];
        const comp = target.findWhere(a => a.key() === 'user-select-screen');
        (comp.getElement() as any).ref({ show: jest.fn() })
        const button = target.findWhere(a => a.key() === 'add-button');
        button.simulate('press');
        expect((instance['selectionModal'] as any).show).toBeCalled();
    });
});
