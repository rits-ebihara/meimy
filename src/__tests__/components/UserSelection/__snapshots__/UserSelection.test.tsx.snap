// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init minimal - no multiple 1`] = `
<Styled(ViewNB)
  style={
    Object {
      "flexDirection": "row",
      "flexWrap": "wrap",
    }
  }
>
  <UserBadge
    badgeColor="#666"
    key="a0001"
    onLongPress={[Function]}
    style={
      Object {
        "margin": 4,
      }
    }
    textColor="#000"
    type="user"
    userId="a0001"
  />
  <Styled(Button)
    icon={true}
    key="add-button"
    onPress={[Function]}
    success={true}
    transparent={true}
  >
    <Styled(Icon)
      name="md-add-circle"
    />
  </Styled(Button)>
  <UserSelectScreen
    filter={
      Object {
        "groups": true,
        "organizations": true,
        "users": true,
      }
    }
    key="user-select-screen"
    onSelect={[Function]}
  />
</Styled(ViewNB)>
`;

exports[`init minimal 1`] = `
<Styled(ViewNB)
  style={
    Object {
      "flexDirection": "row",
      "flexWrap": "wrap",
    }
  }
>
  <UserBadge
    badgeColor="#666"
    key="a0001"
    onLongPress={[Function]}
    style={
      Object {
        "margin": 4,
      }
    }
    textColor="#000"
    type="user"
    userId="a0001"
  />
  <UserBadge
    badgeColor="#666"
    key="a0002"
    onLongPress={[Function]}
    style={
      Object {
        "margin": 4,
      }
    }
    textColor="#000"
    type="user"
    userId="a0002"
  />
  <Styled(Button)
    icon={true}
    key="add-button"
    onPress={[Function]}
    success={true}
    transparent={true}
  >
    <Styled(Icon)
      name="md-add-circle"
    />
  </Styled(Button)>
  <UserSelectScreen
    filter={
      Object {
        "groups": true,
        "organizations": true,
        "users": true,
      }
    }
    key="user-select-screen"
    onSelect={[Function]}
  />
</Styled(ViewNB)>
`;

exports[`init user badge styled 1`] = `
<Styled(ViewNB)
  style={
    Object {
      "flexDirection": "row",
      "flexWrap": "wrap",
    }
  }
>
  <UserBadge
    badgeColor="#900"
    key="a0001"
    onLongPress={[Function]}
    style={
      Object {
        "borderColor": "#090",
        "margin": 4,
      }
    }
    textColor="#ccc"
    type="organization"
    userId="a0001"
  />
  <UserBadge
    badgeColor="#900"
    key="a0002"
    onLongPress={[Function]}
    style={
      Object {
        "borderColor": "#090",
        "margin": 4,
      }
    }
    textColor="#ccc"
    type="organization"
    userId="a0002"
  />
  <UserSelectScreen
    filter={
      Object {
        "groups": true,
        "organizations": true,
        "users": true,
      }
    }
    key="user-select-screen"
    onSelect={[Function]}
  />
</Styled(ViewNB)>
`;
